File: src/day12.py
from collections import defaultdict
from typing import Counter, Dict, Iterator, List

Node = int
START = 0
END = 1
FIRST_BIG = START - 1
FIRST_SMALL = END + 1

node_ids = {"start": START, "end": END}
next_big = FIRST_BIG
next_small = FIRST_SMALL


def lookup(name: str) -> Node:
    global next_big, next_small  # pylint: disable=global-statement
    if name not in node_ids:
        if name.isupper():
            node_ids[name] = next_big
            next_big -= 1
        else:
            node_ids[name] = next_small
            next_small += 1
    return node_ids[name]


def is_big(node: Node) -> bool:
    return node <= FIRST_BIG


# Node = str
# START = "start"
# END = "end"

# def lookup(name: str) -> Node:
#     return name


# def is_big(node: Node) -> bool:
#     return node.isupper()


class Graph:
    """Just an adjacency list."""

    def __init__(self) -> None:
        self._adj: Dict[Node, List[Node]] = defaultdict(list)

    def add_edge(self, a: str, b: str) -> None:
        u = lookup(a)
        v = lookup(b)
        # don't allow backtracking to start
        if v != START:
            self._adj[u].append(v)
        if u != START:
            self._adj[v].append(u)

    def __getitem__(self, u: Node) -> Iterator[Node]:
        return iter(self._adj[u])

    def __iter__(self) -> Iterator[Node]:
        return iter(self._adj)


def parse(lines: List[str]) -> Graph:
    G = Graph()
    for line in lines:
        u, v = line.split("-")
        G.add_edge(u, v)
    return G


def dfs_step(G: Graph, path: List[Node], can_visit_twice: bool = False) -> int:
    """Recursive DFS to count all paths through the graph."""
    count = 0
    node = path[-1]
    for u in G[node]:
        if u == END:
            count += 1
        elif is_big(u) or u not in path:
            count += dfs_step(G, [*path, u], can_visit_twice)
        elif can_visit_twice:
            count += dfs_step(G, [*path, u], False)
    return count


def dfs_iter(G: Graph, can_visit_twice: bool) -> int:
    """Iterative DFS to count all paths through the graph."""
    count = 0
    path = [START]
    iter_stack = [G[path[0]]]
    seen_count = Counter(path)
    while iter_stack:
        neighbors = iter_stack[-1]
        u = next(neighbors, None)
        if u is None:
            # out of neighbors
            iter_stack.pop(-1)
            prev = path.pop(-1)
            if seen_count[prev] == 2 and not is_big(prev):
                can_visit_twice = True
            seen_count[prev] -= 1
        elif u == END:
            count += 1
        else:
            valid = False
            if is_big(u) or seen_count[u] == 0:
                valid = True
            elif can_visit_twice:
                valid = True
                can_visit_twice = False
            if valid:
                seen_count[u] += 1
                path.append(u)
                iter_stack.append(G[u])
    return count


def part_1(lines: List[str]) -> int:
    G = parse(lines)
    # return dfs_step(G, [START])
    return dfs_iter(G, False)


def part_2(lines: List[str]) -> int:
    G = parse(lines)
    # return dfs_step(G, [START], True)
    return dfs_iter(G, True)

File: src/day12.cpp
#include <algorithm>
#include <cassert>
#include <fstream>
#include <iostream>
#include <stack>
#include <string>
#include <unordered_map>
#include <vector>

// typedef std::string Node;

// const Node START = "start";
// const Node END = "end";

// bool is_big(const Node &node) { return std::isupper(node[0]); }

// const Node lookup(const std::string &name) { return name; }

typedef int8_t Node;

constexpr Node START = 0;
constexpr Node END = 1;
constexpr Node FIRST_BIG = START - 1;
constexpr Node FIRST_SMALL = END + 1;

bool is_big(const Node &node) { return node <= FIRST_BIG; }

const Node lookup(const std::string &name) {
  static std::unordered_map<std::string, Node> node_ids{{"start", START},
                                                        {"end", END}};
  if (!node_ids.contains(name)) {
    if (std::isupper(name[0])) {
      static Node next_big = FIRST_BIG;
      node_ids[name] = next_big;
      --next_big;
    } else {
      static Node next_small = FIRST_SMALL;
      node_ids[name] = next_small;
      ++next_small;
    }
  }
  return node_ids.at(name);
}

typedef std::unordered_map<Node, std::vector<Node>> Graph;

void add_edge(Graph &G, const std::string &a, const std::string &b) {
  Node u = lookup(a);
  Node v = lookup(b);
  if (v != START)
    G[u].push_back(v);
  if (u != START)
    G[v].push_back(u);
}

Graph parse(std::ifstream input) {
  Graph G{};
  std::string line, u, v;
  while (std::getline(input, line)) {
    size_t sep_pos = line.find_first_of('-');
    u = line.substr(0, sep_pos);
    v = line.substr(sep_pos + 1, line.size() - sep_pos - 1);
    add_edge(G, u, v);
  }
  return G;
}

int dfs_step(const Graph &G, const std::vector<Node> &path,
             bool can_visit_twice) {
  int count = 0;
  const Node node = path.back();
  for (auto &&u : G.at(node)) {
    if (u == END) {
      ++count;
      continue;
    }
    bool valid = false, new_can_visit = can_visit_twice;
    if (is_big(u) || std::find(path.cbegin(), path.cend(), u) == path.cend()) {
      valid = true;
    } else if (can_visit_twice) {
      valid = true;
      new_can_visit = false;
    }
    if (valid) {
      std::vector<Node> new_path{path};
      new_path.push_back(u);
      count += dfs_step(G, new_path, new_can_visit);
    }
  }
  return count;
}

#define stack_entry(n)                                                         \
  { (n), G.at((n)).cbegin(), G.at((n)).cend() }
int dfs_iter(const Graph &G, bool can_visit_twice) {
  int count = 0;
  // std::stack<Node> path{{START}};
  std::stack<std::tuple<Node, Graph::mapped_type::const_iterator,
                        const Graph::mapped_type::const_iterator>>
      stack{{stack_entry(START)}};
  std::unordered_map<Node, int> seen_count{{START, 1}};
  while (!stack.empty()) {
    auto &[prev, iter, end] = stack.top();
    if (iter == end) {
      // out of neighbors
      if (seen_count[prev] == 2 && !is_big(prev)) {
        can_visit_twice = true;
      }
      --seen_count[prev];
      stack.pop();
    } else {
      Node u = *(iter++);
      if (u == END) {
        ++count;
      } else {
        bool valid = false;
        if (is_big(u) || seen_count[u] == 0) {
          valid = true;
        } else if (can_visit_twice) {
          valid = true;
          can_visit_twice = false;
        }
        if (valid) {
          ++seen_count[u];
          stack.push(stack_entry(u));
        }
      }
    }
  }
  return count;
}
#undef stack_entry

int main(int argc, char **argv) {
  std::string input_path;
  std::vector<std::string> args{};
  for (int i = 0; i < argc; ++i) {
    args.push_back(argv[i]);
  }
  if (args.size() < 2) {
    input_path = "input/day12.txt";
  } else if (args[1] == "-e") {
    input_path = "example/day12.txt";
  } else if (args[1] == "-i") {
    input_path = args[2];
  }

  Graph G = parse(std::ifstream{input_path});

  // std::cout << "Part 1: " << dfs_step(G, {START}, false) << std::endl;
  // std::cout << "Part 2: " << dfs_step(G, {START}, true) << std::endl;
  // std::cout << "Part 1: " << dfs_iter(G, false) << std::endl;
  std::cout << "Part 2: " << dfs_iter(G, true) << std::endl;
  return 0;
}
